### PR TITLE
feat: Added feature to run rendering in synchronous mode. 

### PR DESCRIPTION
**Pull request by fritzduchardt**
This makes debugging easier and takes load of slow machines. It has proven useful in the past.

### CODE REVIEW

File: internal/myks/util_test.go [hunk 1]
```diff
@@ -239,6 +239,9 @@ func Test_runCmd(t *testing.T) {
 }
 
 func Test_runYttWithFilesAndStdin(t *testing.T) {
+	if os.Getenv("CI") == "true" {
+		t.Skip("Skipping in pipeline since ytt is not installed")
+	}
```
Comment (Zebradil) [inline internal/myks/util_test.go:244]: We actually can just install all the needed tools.

**Review by Zebradil — COMMENTED** [review]
What do you think about a flag to control the level of parallelism instead of just switching it on and off?

- `--async=0` to disable
- `--async=1` to enable
- `--async=8` to run maximum of 8 parallel threads

For now, we can skip implementing the last one, but I would like to settle with the flag name.

File: internal/myks/util_test.go (hunk 1 above)
Comment (fritzduchardt) [reply internal/myks/util_test.go:244]: Fair point, will look into that.

Comment (Zebradil) [reply internal/myks/util_test.go:244]: Here is an example of how to install required carvel tools in CI: https://github.com/mykso/setup-action/blob/2966421d42a27cde1881282ba85a089c6d09c030/action.yml#L31-L38

Comment (fritzduchardt) [reply internal/myks/util_test.go]: Done

File: .github/workflows/job-test.yml [hunk 2]
```diff
@@ -14,6 +14,14 @@ jobs:
       - uses: actions/setup-go@v4
         with:
           go-version: "1.20"
+      - uses: actions/setup-go@v4
+        with:
+          go-version: "1.20"
```
Comment (kbudde) [inline .github/workflows/job-test.yml:19]: Why do you set up go twice?

File: internal/myks/util.go [hunk 3]
```diff
@@ -107,19 +123,20 @@ func copyFileSystemToPath(source fs.FS, sourcePath string, destinationPath strin
...
-			defer srcFile.Close()
+			defer saveClose(srcFile)
 
 			// Create the destination file
 			dstFile, err := os.Create(destination)
 			if err != nil {
 				return err
 			}
-			defer dstFile.Close()
+			defer saveClose(dstFile)
```
Comment (kbudde) [inline internal/myks/util.go:139]: Error is logged but not returned. 
Named result values are your friend :) https://blog.learngoprogramming.com/5-gotchas-of-defer-in-go-golang-part-iii-36a1ab3d6ef1

File: .github/workflows/job-test.yml (hunk 2 above)
Comment (fritzduchardt) [reply .github/workflows/job-test.yml:19]: Better save than sorry ;) 

Good spot

File: internal/myks/util.go (hunk 3 above)
Comment (fritzduchardt) [reply internal/myks/util.go:143]: Good point, using them now, but got to say, they are NOT my friends looking at the complexity of the resulting code:

```
saveClose := func(srcFile fs.File) {
		closeErr := srcFile.Close()
		if closeErr != nil {
			if err == nil {
				err = closeErr
			} else {
				err = errors.Join(err, closeErr)
			}
		}
	}
```

Comment (kbudde) [reply internal/myks/util.go:143]: In this case, what about:
```go
saveClose := func(srcFile fs.File) {
	closeErr := srcFile.Close()
	err = errors.Join(err, closeErr)
}
```
errors.Join is ignoring nil errors. If all errors are nil it'll return nil.
Downside you always have to unwrap the error.

File: cmd/root.go [hunk 4]
```diff
@@ -68,6 +69,7 @@ var rootCmd = &cobra.Command{
 func init() {
 	cobra.OnInitialize(initConfig)
 	rootCmd.PersistentFlags().StringP("log-level", "l", "info", "Set the logging level")
+	rootCmd.PersistentFlags().IntVarP(&asyncLevel, "async", "a", 1, "Run asynchronously. Activated by default. Set to 0 to deactivate.")
```
Comment (kbudde) [inline cmd/root.go:72]: Now as async controls the number of parallel routines, I would propose to set some better default.
1 is the same as 0 just using different code. 
What about 4?

File: internal/myks/util.go [hunk 5]
```diff
@@ -34,52 +36,58 @@ func reductSecrets(args []string) []string {
...
+	default:
+		return fmt.Errorf("collection must be a slice, array or map, got %s", value.Kind())
 	}
 
-	errChan := make(chan error, len(items))
-	defer close(errChan)
+	// run async
+	if asyncLevel > 0 {
+		var eg errgroup.Group
+		semaphore := make(chan struct{}, asyncLevel)
```
Comment (kbudde) [inline internal/myks/util.go:59]: Could be replaced by `eg.SetLimit(asyncLevel)`. No semaphore needed in this case :) 
https://pkg.go.dev/golang.org/x/sync/errgroup#Group.SetLimit

File: cmd/root.go (hunk 4 above)
Comment (Zebradil) [reply cmd/root.go:72]: What about inverting the logic a little bit and doing the following:

- async level is `N` means that maximum `N` applications are being processed in parallel,
- async level is `0` means the limit is not set, any number of applications can be processed in parallel.

In this case, the default might be `0`. I see it hard to set a viable default value. Moreover, for different operations (sync, render) the "ideal" number would differ.

**Comment by kbudde** [comment]
@fritzduchardt , @Zebradil I added one commit. I think it can be merged now.

**Pull request merged**